uc-crux-llvm/src/UCCrux/LLVM/FullType/CrucibleType.hs

```diff
@@ -0,0 +1,275 @@
+Copyright        : (c) Galois, Inc 2021
+License          : BSD3
+Maintainer       : Langston Barrett <langston@galois.com>
+Stability        : provisional
+-}
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE ImplicitParams #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE PolyKinds #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE TypeOperators #-}
+
+module UCCrux.LLVM.FullType.CrucibleType
+  ( toCrucibleType,
+
+    -- * Assignments
+    SomeAssign (..),
+    SomeIndex (..),
+    assignmentToFullType,
+    SomeAssign' (..),
+    assignmentToCrucibleType,
+    assignmentToCrucibleTypeA,
+    translateIndex,
+    generateM,
+    generate,
+  )
+where
+
+{- ORMOLU_DISABLE -}
+import           Prelude hiding (unzip)
+
+import           Data.Coerce (coerce)
+import           Data.Functor ((<&>))
+import           Data.Functor.Const (Const(Const, getConst))
+import           Data.Functor.Identity (Identity(Identity, runIdentity))
+import           Data.Functor.Product (Product(Pair))
+import           Control.Monad.State (runStateT)
+
+import qualified Text.LLVM.AST as L
+
+import qualified Data.Parameterized.Context as Ctx
+import           Data.Parameterized.NatRepr
+import           Data.Parameterized.Some (Some(Some))
+
+import qualified Lang.Crucible.Types as CrucibleTypes hiding ((::>))
+
+import qualified Lang.Crucible.LLVM.MemModel as LLVMMem
+import           Lang.Crucible.LLVM.MemType (MemType(..))
+import           Lang.Crucible.LLVM.TypeContext (TypeContext)
+
+import           Crux.LLVM.Overrides (ArchOk)
+import           UCCrux.LLVM.Errors.Panic (panic)
+import           UCCrux.LLVM.FullType.Type
+import           UCCrux.LLVM.FullType.MemType (toFullTypeM)
+import           UCCrux.LLVM.FullType.ModuleTypes (ModuleTypes, makeModuleTypes)
+{- ORMOLU_ENABLE -}
+
+-- | c.f. @llvmTypeToRepr@
+toCrucibleType ::
+  ArchOk arch =>
+  proxy arch ->
+  FullTypeRepr m ft ->
+  CrucibleTypes.TypeRepr (ToCrucibleType arch ft)
+toCrucibleType proxy =
+  \case
+    FTIntRepr natRepr -> LLVMMem.LLVMPointerRepr natRepr
+    FTPtrRepr _ -> LLVMMem.LLVMPointerRepr ?ptrWidth
+    FTVoidFuncPtrRepr _argTypeReprs -> LLVMMem.LLVMPointerRepr ?ptrWidth
+    FTNonVoidFuncPtrRepr _retTypeRepr _argTypeReprs -> LLVMMem.LLVMPointerRepr ?ptrWidth
+    FTOpaquePtrRepr _aliasName -> LLVMMem.LLVMPointerRepr ?ptrWidth
+    FTFloatRepr floatInfo -> CrucibleTypes.FloatRepr floatInfo
+    FTArrayRepr _natRepr fullTypeRepr ->
+      CrucibleTypes.VectorRepr (toCrucibleType proxy fullTypeRepr)
+    FTStructRepr _ typeReprs ->
+      case assignmentToCrucibleType proxy typeReprs of
+        SomeAssign' ctReprs Refl _ -> CrucibleTypes.StructRepr ctReprs
+
+data SomeAssign arch crucibleTypes = forall m fullTypes.
+  SomeAssign
+  { saFullTypes :: Ctx.Assignment (FullTypeRepr m) fullTypes,
+    saModuleTypes :: ModuleTypes m,
+    saProof :: MapToCrucibleType arch fullTypes :~: crucibleTypes
+  }
```

Comment:
Instead of an explicit `Refl` for the proof term, you could push that equality into the existentially-quantified constraint of `SomeAssign`. You would then recover it whenever you pattern match on the `SomeAssign`

Reply:
https://github.com/GaloisInc/crucible/issues/693